velox/docs/ext/spark.py

```diff
@@ -0,0 +1,1022 @@
+            return name + parens
+        if config.toc_object_entries_show_parents == "all":
+            return ".".join(parents + [name + parens])
+        return ""
+
+
+class SparkFunction(SparkObject):
+    """Description of a function."""
+
+    option_spec: OptionSpec = SparkObject.option_spec.copy()
+    option_spec.update(
+        {
+            "async": directives.flag,
+        }
+    )
+
+    def get_signature_prefix(self, sig: str) -> list[nodes.Node]:
+        if "async" in self.options:
+            return [addnodes.desc_sig_keyword("", "async"), addnodes.desc_sig_space()]
+        else:
+            return []
+
+    def needs_arglist(self) -> bool:
+        return True
+
+    def add_target_and_index(
+        self, name_cls: tuple[str, str], sig: str, signode: desc_signature
+    ) -> None:
+        super().add_target_and_index(name_cls, sig, signode)
+        if "noindexentry" not in self.options:
+            modname = self.options.get(
+                "module", self.env.ref_context.get("spark:module")
+            )
+            node_id = signode["ids"][0]
+
+            name, cls = name_cls
+            if modname:
+                text = _("%s() (in module %s)") % (name, modname)
+                self.indexnode["entries"].append(("single", text, node_id, "", None))
+            else:
+                text = f'{pairindextypes["builtin"]}; {name}()'
+                self.indexnode["entries"].append(("pair", text, node_id, "", None))
+
+    def get_index_text(self, modname: str, name_cls: tuple[str, str]) -> str | None:
+        # add index in own add_target_and_index() instead.
+        return None
+
+
+class SparkXRefRole(XRefRole):
+    def process_link(
+        self,
+        env: BuildEnvironment,
+        refnode: Element,
+        has_explicit_title: bool,
+        title: str,
+        target: str,
+    ) -> tuple[str, str]:
+        refnode["spark:module"] = env.ref_context.get("spark:module")
+        refnode["spark:class"] = env.ref_context.get("spark:class")
+        if not has_explicit_title:
+            title = title.lstrip(".")  # only has a meaning for the target
+            target = target.lstrip("~")  # only has a meaning for the title
+            # if the first character is a tilde, don't display the module/class
+            # parts of the contents
+            if title[0:1] == "~":
+                title = title[1:]
+                dot = title.rfind(".")
+                if dot != -1:
+                    title = title[dot + 1 :]
+        # if the first character is a dot, search more specific namespaces first
+        # else search builtins first
+        if target[0:1] == ".":
+            target = target[1:]
+            refnode["refspecific"] = True
+        return title, target
+
+
+class SparkModuleIndex(Index):
+    """
+    Index subclass to provide the Python module index.
+    """
+
+    name = "modindex"
+    localname = _("Python Module Index")
```

Comment:
All python were replaced by spark except `subtype="python"`.  we can not let `subtype="python"` change to `subtype="spark"`, because the link will fail.